### PR TITLE
monitor: resolve paths arguments in client

### PR DESCRIPTION
**Pull request by ktock**
Following up for https://github.com/docker/buildx/pull/1296#issuecomment-1410470087

Quoted from @jedevc 's comment (https://github.com/docker/buildx/pull/1296#issuecomment-1410470087):

> ```
> BUILDX_EXPERIMENTAL=1 buildx build . --builder=dev --invoke=sh
> INFO: connecting to buildx server
> [+] Building 0.0s (2/2) FINISHED                                                                                                                                  
>  => [internal] load build definition from Dockerfile                                                                                                         0.0s
>  => => transferring dockerfile: 2B                                                                                                                           0.0s
>  => [internal] load .dockerignore                                                                                                                            0.0s
>  => => transferring context: 2B                                                                                                                              0.0s
> ERROR: failed to build: rpc error: code = Unknown desc = failed to solve: failed to read dockerfile: open /tmp/buildkit-mount3665150422/Dockerfile: no such file or directory
> ```

This commit allows using relative paths on `buildx build` arguments by fixing the controller client to resolve paths on the client side.

### CODE REVIEW

**Comment by jedevc** [comment]
Hm, I'm fairly sure we should look for a better way to go about this :thinking:

As is, this looks *really* tricky to maintain over time, and I think it's best to avoid modifying the user-provided options wherever possible. I think we need to push down the file resolution to the components that actually open the files, though those are scattered in BuildKit's client code as well as in Buildx.

For example, in Buildkit, we need to handle it here:

https://github.com/docker/buildx/blob/6506166f02c3c856507a1b1708e1cd6fac2bf614/vendor/github.com/moby/buildkit/client/solve.go#L193-L201

To get the information of the right path through - I think we need to add a current working directory field to the `SolveOpt`. Then that gives us the ability to handle the paths used in the cache importers/exporters, as well as the dockerfile and contexts.

For things like the metadata file, and secrets, we need to handle those buildx-side - so we'd need to add the same working directory property to the `controllerapi.BuildOptions`.

I think this would work better as a long-term approach? The buildkit-side patches shouldn't be too complex, and should just affect the client, so we wouldn't need new capabilities or anything.

@tonistiigi are we missing an easier option here?

**Comment by jedevc** [comment]
With #1614 merged, @ktock, could you rebase this with those most recent changes?

**Comment by ktock** [comment]
@jedevc Thank you for working on #1614. Rebased this PR on that commit.

File: commands/build.go [hunk 1]
```diff
@@ -649,3 +657,154 @@ func dockerUlimitToControllerUlimit(u *dockeropts.UlimitOpt) *controllerapi.Ulim
 	}
 	return &controllerapi.UlimitOpt{Values: values}
 }
+
+// resolvePaths resolves all paths contained in controllerapi.BuildOptions
+// and replaces them to absolute paths.
+func resolvePaths(options *controllerapi.BuildOptions) (_ *controllerapi.BuildOptions, err error) {
+	if options.ContextPath != "-" {
+		options.ContextPath, err = absIfLocal(options.ContextPath)
```
Comment (jedevc) [inline commands/build.go:665]: I'm not sure if we should do `os.Stat` here as part of `absIfLocal`? I think we should just do `filepath.Abs`, and similarly below.

We know at this point it's definitely a path, and if the file doesn't exist, we should let the controller throw the error (which may or may not be a file not found error, we shouldn't assume the implementation at this point).

Comment (ktock) [reply commands/build.go]: Thank you for the comment. Yes, we don't need os.Stat here because the options are already strongly typed. Fixed the patch.

Comment (jedevc) [reply commands/build.go]: I just pushed a commit on top to handle the new oci-layout named context (which can also be a path). Does that seem good to you? If so, I think this is ready-to-go :tada:

**Review by jedevc — APPROVED** [review]

File: commands/build.go (hunk 1 above)
Comment (ktock) [reply commands/build.go]: Thank you for the following-up! LGTM.

**Pull request merged**